src/OpenTelemetry.Instrumentation.EventCounters/EventCountersMetrics.cs

```diff
@@ -0,0 +1,100 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.Metrics;
+using System.Diagnostics.Tracing;
+using System.Reflection;
+
+namespace OpenTelemetry.Instrumentation.EventCounters;
+
+/// <summary>
+/// .NET EventCounters Instrumentation.
+/// </summary>
+internal class EventCountersMetrics : EventListener
+{
+    internal static readonly AssemblyName AssemblyName = typeof(EventCountersMetrics).Assembly.GetName();
+    internal static readonly Meter MeterInstance = new(AssemblyName.Name, AssemblyName.Version.ToString());
+
+    private readonly EventCountersInstrumentationOptions options;
+    private readonly ConcurrentBag<EventSource> preInitEventSources = new();
+    private readonly ConcurrentDictionary<Tuple<string, string>, Instrument> instruments = new();
+    private readonly ConcurrentDictionary<Tuple<string, string>, double> values = new();
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="EventCountersMetrics"/> class.
+    /// </summary>
+    /// <param name="options">The options to define the metrics.</param>
+    public EventCountersMetrics(EventCountersInstrumentationOptions options)
+    {
+        this.options = options;
+
+        while (this.preInitEventSources.TryTake(out EventSource source))
+        {
+            if (this.options.ShouldListenToSource(source.Name))
+            {
+                this.EnableEvents(source, EventLevel.LogAlways, EventKeywords.None, GetEnableEventsArguments(this.options));
+            }
+        }
+    }
+
+    /// <inheritdoc />
+    protected override void OnEventSourceCreated(EventSource source)
+    {
+        if (this.options == null)
+        {
+            this.preInitEventSources.Add(source);
+        }
+        else if (this.options.ShouldListenToSource(source.Name))
+        {
+            this.EnableEvents(source, EventLevel.LogAlways, EventKeywords.None, GetEnableEventsArguments(this.options));
+        }
+    }
+
+    /// <inheritdoc />
+    protected override void OnEventWritten(EventWrittenEventArgs eventData)
+    {
+        if (this.options == null)
+        {
+            return;
+        }
+
+        var payload = eventData.Payload[0] as IDictionary<string, object>;
+        var name = payload["Name"].ToString();
+        var isGauge = payload.ContainsKey("Mean");
+        Tuple<string, string> metricKey = new(eventData.EventSource.Name, name);
+
+        if (isGauge)
+        {
+            this.values[metricKey] = Convert.ToDouble(payload["Mean"]);
+        }
+        else
+        {
+            _ = this.values.TryAdd(metricKey, 0);
+            this.values[metricKey] += Convert.ToDouble(payload["Increment"]);
+        }
+
+        var instrumentName = $"{eventData.EventSource.Name}.{name}";
+        _ = this.instruments.TryAdd(
```

Comment:
This is going to create an instrument on every `OnEventWritten` call. Use `ConcurrentDictionary.ContainsKey` instead.

Reply:
It will only create an instrument if `!this.instruments.ContainsKey(metricKey)` is `false`
I've just pushed some updated code though so please check again

Reply:
We have the `Meter.CreateObservableCounter`/ `Meter.CreateObservableGauge` in the `TryAdd` which would create an instrument on every `TryAdd` call.

I created a simple Benchmark with simpler data types: `int` and `string` to verify this behavior. It indeed would create the value every time as we have allocation here:

| Method |     Mean |    Error |   StdDev |   Gen0 | Allocated |
|------- |---------:|---------:|---------:|-------:|----------:|
| TryAdd | 61.76 ns | 1.745 ns | 5.144 ns | 0.0357 |     224 B |

<details>

<summary>Expand for code</summary>

```csharp
using System.Collections.Concurrent;
using BenchmarkDotNet.Attributes;

namespace Benchmarks
{
    [MemoryDiagnoser]
    public class ConcurrentDictionaryBenchmarks
    {
        private readonly ConcurrentDictionary<int, string> dictionary = new();

        public ConcurrentDictionaryBenchmarks()
        {
            _ = this.dictionary.TryAdd(100, new string('A', 100));
        }

        [Benchmark]
        public void TryAdd()
        {
            _ = this.dictionary.TryAdd(100, new string('A', 100));
        }
    }
}
```

Reply:
Yes I agree with this. In the surrounding scope I do have a `if (!this.instruments.ContainsKey(metricKey))` check though which might have been hidden in the GitHub code preview?